Support erofs overlaydisk

schaefi opened the pull request:
So this change is many fold and split into several commits. The most important changes summarizes into

**Update test-image-overlayroot integration test**
    
Switch to erofs for overlay testing. Additionally split the build into two profiles. The first one just builds a simple overlayroot oem disk based on erofs. The second one adds a veritysetup layer and configures the systemd-veritysetup-generator for use in dracut.

**Add support for selecting the overlay read-only fs**
    
Add new overlayroot_readonly_filesystem attribute which allows to select for either squashfs or erofs as the read-only filesystem  in an OEM overlay disk setup.

**Add documentation for new attribute**
    
Add details how to use the new overlayroot_readonly_filesystem attribute

**Fixed root setup for verity overlay disk**
    
When building an image with overlayroot set to true and activated verity data, the root= parameter must be set to root=overlay:MAPPER=verityroot instead of the standard  overlay:PARTUUID mapping.


This Fixes #2799

schaefi commented:
I will zip a Staging build including this change and let the integration test run. This will produce an image for testing. Until then the no-merge label is set

schaefi commented:
Image builds can be found at

* https://build.opensuse.org/package/show/Virtualization:Appliances:Images:Testing_x86:tumbleweed/test-image-overlayroot

schaefi commented:
I still had no luck with the systemd verity generator and added the simple parsing of `/etc/veritytab` to the existing `kiwi-verity` dracut module. This allows for a simple alternative and works nicely. Once the systemd verity generator works we only need to remove the `dracut-kiwi-verity` package from the list and everything should continue in the same way as before.

With the current setup my integration test image comes up with the following layout

```
localhost:~ # lsblk 
NAME           MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINTS
fd0              2:0    1    4K  0 disk  
sda              8:0    0    4G  0 disk  
├─sda1           8:1    0   20M  0 part  /boot/efi
├─sda2           8:2    0  300M  0 part  /boot
├─sda3           8:3    0  909M  0 part  
│ └─verityroot 254:0    0  792M  1 crypt /run/overlay/rootfsbase
└─sda4           8:4    0  2.8G  0 part  /run/overlay/overlayfs
sr0             11:0    1 1024M  0 rom   

localhost:~ # mount | grep erofs
/dev/mapper/verityroot on /run/overlay/rootfsbase type erofs (ro,relatime,user_xattr,acl,cache_strategy=readaround)
```

Except for secure boot this should be pretty close to the desired behavior

schaefi commented:
Questions to discuss maybe

- [ ] Use systemd-boot instead of grub in the integration test ?
- [ ] Use another filesystem for the write partition, currently it's ext3, maybe better xfs ?

Conan-Kudo commented:
We should have both grub and sd-boot, and probably using xfs and btrfs as those are the expected filesystems to be used with this.

schaefi commented:
I updated the integration test and moved to systemd-boot as well as enabled secure boot and also dropped the extra boot partition. This aligns now nicely to the following structure

```
localhost:~ # lsblk 
NAME           MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
fd0              2:0    1     4K  0 disk  
sda              8:0    0     4G  0 disk  
├─sda1           8:1    0   200M  0 part  /boot/efi
├─sda2           8:2    0   900M  0 part  
│ └─verityroot 254:0    0 853.1M  1 crypt /run/overlay/rootfsbase
└─sda3           8:3    0   2.9G  0 part  /run/overlay/overlayfs
sr0             11:0    1  1024M  0 rom   

```

and bootctl says 

```
System:
      Firmware: UEFI 2.70 (EDK II 1.00)
 Firmware Arch: x64
   Secure Boot: disabled (setup)
  TPM2 Support: no
  Measured UKI: no
  Boot into FW: supported

Current Boot Loader:
      Product: systemd-boot 257.5+suse.8.gc10a66fb4d
     Features: ✓ Boot counting
               ✓ Menu timeout control
               ✓ One-shot menu timeout control
               ✓ Default entry control
               ✓ One-shot entry control
               ✓ Support for XBOOTLDR partition
               ✓ Support for passing random seed to OS
               ✓ Load drop-in drivers
               ✓ Support Type #1 sort-key field
               ✓ Support @saved pseudo-entry
               ✓ Support Type #1 devicetree field
               ✓ Enroll SecureBoot keys
               ✓ Retain SHIM protocols
               ✓ Menu can be disabled
               ✓ Multi-Profile UKIs are supported
               ✓ Boot loader set partition information
    Partition: /dev/disk/by-partuuid/f7302a8b-4790-4e84-89dd-0a87ee3b7867
       Loader: └─/EFI/BOOT/BOOTX64.EFI

```

schaefi commented:
I need to polish the tests and then another commit to this PR will follow

schaefi commented:
> We should have both grub and sd-boot, and probably using xfs and btrfs as those are the expected filesystems to be used with this.

@Conan-Kudo agreed. I suggest to add the grub based overlay test as an extra integration test. Just because adding more profiles to the existing overlay integration test makes it harder to read and maintain the description. I will keep this in mind as a followup once this PR gets merged

schaefi commented:
I pushed a package to Staging. Once all integration tests builds without issues, this PR is ready for review/merge. Thanks

schaefi commented:
> I'd rather have more tests added with this PR because we've been bitten by not knowing these things properly before.

I added another one for grub instead of sdboot and btrfs instead of xfs. I think the current number of tests for the overlayroot oem disk is now sufficient

schaefi commented:
All integration tests passing
